internal/guest/amdsev/report.go

```diff
@@ -0,0 +1,266 @@
+	ReportIDMA       [32]byte
+	ReportTCB        uint64
+	Reserved2        [24]byte
+	ChipID           [64]byte
+	CommittedSVN     [8]byte
+	CommittedVersion [8]byte
+	LaunchSVN        [8]byte
+	Reserved3        [168]byte
+	Signature        [512]byte
+}
+
+func (sr *snpReport) report() Report {
+	return Report{
+		Version:          sr.Version,
+		GuestSVN:         sr.GuestSVN,
+		Policy:           sr.Policy,
+		FamilyID:         hex.EncodeToString(mirrorBytes(sr.FamilyID[:])[:]),
+		ImageID:          hex.EncodeToString(mirrorBytes(sr.ImageID[:])[:]),
+		VMPL:             sr.VMPL,
+		SignatureAlgo:    sr.SignatureAlgo,
+		PlatformVersion:  sr.PlatformVersion,
+		PlatformInfo:     sr.PlatformInfo,
+		AuthorKeyEn:      sr.AuthorKeyEn,
+		ReportData:       hex.EncodeToString(sr.ReportData[:]),
+		Measurement:      hex.EncodeToString(sr.Measurement[:]),
+		HostData:         hex.EncodeToString(sr.HostData[:]),
+		IDKeyDigest:      hex.EncodeToString(sr.IDKeyDigest[:]),
+		AuthorKeyDigest:  hex.EncodeToString(sr.AuthorKeyDigest[:]),
+		ReportID:         hex.EncodeToString(sr.ReportID[:]),
+		ReportIDMA:       hex.EncodeToString(sr.ReportIDMA[:]),
+		ReportTCB:        sr.ReportTCB,
+		ChipID:           hex.EncodeToString(sr.ChipID[:]),
+		CommittedSVN:     hex.EncodeToString(sr.CommittedSVN[:]),
+		CommittedVersion: hex.EncodeToString(sr.CommittedVersion[:]),
+		LaunchSVN:        hex.EncodeToString(sr.LaunchSVN[:]),
+		Signature:        hex.EncodeToString(sr.Signature[:]),
+	}
+}
+
+// msgReportResponse is the attestation response struct
+// https://www.amd.com/system/files/TechDocs/56860.pdf
+// MSG_REPORT_RSP Table 23.
+// NOTE: msgReportResponse.Report is a byte slice, to have the original
+//     response in bytes. The conversion to internal struct happens inside
+//     convertRawReport.
+// NOTE: the additional 64 bytes are reserved, without them, the ioctl fails.
+type msgReportResponse struct {
+	Status     uint32
+	ReportSize uint32
+	reserved1  [24]byte
+	Report     [1184]byte
+	reserved2  [64]byte // padding to the size of SEV_SNP_REPORT_RSP_BUF_SZ (i.e., 1280 bytes)
+}
+
+// FetchRawSNPReport returns attestation report bytes.
+func FetchRawSNPReport(reportData string) ([]byte, error) {
+	f, err := os.OpenFile("/dev/sev", os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		f.Close()
+	}()
+
+	msgReportIn := msgReportRequest{}
+	msgReportOut := msgReportResponse{}
+
+	if reportData != "" {
+		if len(reportData) > len(msgReportIn.ReportData) {
+			return nil, fmt.Errorf("reportData too large: %s", reportData)
+		}
+		rd, err := hex.DecodeString(reportData)
+		if err != nil {
+			return nil, err
+		}
+		copy(msgReportIn.ReportData[:], rd[:])
+	}
+
+	payload := &SevSNPGuestRequest{
+		RequestMsgType:  MsgReportRequest,
+		ResponseMsgType: MsgReportResponse,
+		MsgVersion:      1,
+		RequestLength:   uint16(unsafe.Sizeof(msgReportIn)),
+		RequestUAddr:    uint64(uintptr(unsafe.Pointer(&msgReportIn))),
```

Comment:
These fields where we are storing addresses of Go values as `uintptr` are also risky. A couple of options:

- Allocate memory outside of Go and copy the data, then pass the address of the allocated memory instead. On Windows this isn't too bad, but not sure if it's easy on Linux.
- Store a `unsafe.Pointer` directly rather than a `uintptr`.

The second option seems much easier.